### ISSUE TITLE
Recognition of handwritten text

### ISSUE THREAD

**Issue by rolandlo** [xournalpp/xournalpp#2152]
**Is your feature request related to a problem? Please describe.**
It would be nice to have a feature to extract the handwritten text from a Xournal++ document, so you can reuse it in a text editor. 

**Describe the solution you'd like**
I would like to have a Xournal++ plugin, in which you can select pages (or the whole document) and get the text extracted from the selected pages and save it to a .txt file or copy it to clipboard.

**Describe alternatives you've considered**
Instead of using a Xournal++ plugin we might create an external tool that takes a xopp-file (and page numbers) as input and returns the extracted text as output.

**Additional context**
This idea has also been discussed in #51 and #1953. 
Using the google API to recognize text the task might be much easier than expected. See [this online demo](https://www.chenyuho.com/project/handwritingjs/) and note that the input sent to the google API could be very easily produced from the list of strokes that Xournal++ produces, see [this description](https://github.com/ChenYuHo/handwriting.js#capture-trace).

**Comment by rolandlo** [comment]
Got a first working version, already. Currently upon activating the plugin, the recognized text (a list of possibilities, sorted by likelyhood) is written to the console. See 
![FirstImpression](https://user-images.githubusercontent.com/40485433/89757082-8fe88380-dae4-11ea-9894-cf0cd083840c.png)

A PR will follow.

**Comment by atticus-sullivan** [comment]
Well this looks pretty amazing. The only little concern that I'm having, is that handwriting data is quite personal and I personally would like to know more about how this works (is the data used to improve the AI? Will the selection the user makes which was the correct guess be sent back?).

So I think this is a really nice feature (searching ones own Notes is awesome), but in my opinion this should be (if possible) something like a plugin or something the user can turn on and off, if he/she has concerns.

(Don't get me wrong this is amazing, but I think one should be careful with this)

**Comment by rolandlo** [comment]
@atticus-sullivan 
You are right about privacy concerns. With this plugin the handwritten data (coordinates of the points making up the strokes) is sent to the google server and processed there. We have no control what google is using that data for. 
It is designed as a Lua Plugin and deactivated by default. You first have to activate it and only after clicking on the name of the plugin the data is sent to the google server. In a more refined version you would first get a dialog where you can choose the pages/layers which should get recognized and then you have to click a button to send the data to the google server. You would see a clear warning there.
When you select the correct guess no data is sent to the google server.

**Comment by rolandlo** [comment]
@atticus-sullivan Currently the UI of my Lua Plugin looks like this:
![grafik](https://user-images.githubusercontent.com/40485433/90023186-b90b3e80-dcb3-11ea-9336-3faf97e0f5bc.png)

Is that enough of a warning in your oppinion?

**Comment by atticus-sullivan** [comment]
I'm no lawyer, but I think this should be enough (don't know maybe a link to the terms which one has to accept would be good (even though no one or only a few people will use it) if there is one (explicitly for this service)).

I didn't thought of such a detailed setting, selecting even only the layer which should be recognized. So I'm already positively impressed ;)

Still, I think the option "Number of returns" should be explained in more detail.

**Comment by rolandlo** [comment]
Well, I am also no lawyer. This google API lacks official documentation (as far as I know). I don't know if any terms are in use by making a https-request to a server.
About the "number of returns": That's the number of different recognized texts you get as alternatives (maximally). Maybe it could be named more understandably. If you have a suggestion let me know.

**Comment by rolandlo** [comment]
A draft PR is provided by #2176

**Comment by rolandlo** [comment]
Tesseract is suboptimal for the task at hand. It is designed for off-line text recognition, where you only process an image and have no information, how it was written. 
Our data (strokes) is suited for on-line recognition. The approach with the google server is on-line recognition. Citing Rao and Aditya:

> Handwriting data is converted to digital form either by scanning
> the writing on paper or by writing with a special pen on an
> electronic surface. The two approaches are distinguished as off-
> line and on-line handwriting, respectively. In the on-line case, the
> two-dimensional coordinates of successive points of the writing
> as a function of time are stored in order. In the off-line case, only
> the completed writing is available as an image.
> The recognition rates reported are much
> higher for the on-line case in comparison with the off-line case.
> Off-line systems are less accurate than on-line systems.

**Comment by chron-isch** [comment]
I love the work, but to be honest, I don't really have any usecase for simply exporting text from a Xournalpp document, I guess it could be useful in certain cases.

But what I would consider really useful is having that feature while exporting to PDF, you could save the machine readable text along with the actual handwritten notes like OCR usually does it.

**Comment by tyleroconnell** [comment]
Would solving this issue help move Xournalapp towards making handwritting within documents searchable? I agree with @chron-isch that exporting text on its own is a bit niche but surely search would be a widely used feature.

**Comment by rolandlo** [comment]
@chron-isch @tyleroconnell 
I agree with both of you. Unfortunately going from the current plugin to searchable handwritten notes is a pretty long way to go. The google server only returns the recognized text, but doesn't give any clues about which strokes correspond to which letters. This could maybe be achieved for text written nicely on lines, but for annotations spread all across the document this would probably be almost impossible. 
So don't expect too much of the current approach. It's a little demo that helps me to figure out how to extend the Lua plugin API and what can be achieved with Lua plugins in Xournal++, but it most likely won't lead to searchable handwritten notes.

**Comment by All3xJ** [comment]
Great project, any news?

**Comment by All3xJ** [comment]
It would be great to do what Goodnotes app does: OCR the handwritten text, to make it searchble. This would make Xournalpp incredible and unbeatable. I wish I could help to develop this, but I don't think I have the right knowledge. Have you any ideas on this? Does this issue report ever bringed to a result?

**Comment by saikm** [comment]
If OCR text was saved into the PDF and was searchable that would be a very useful feature. Right now there is no note taking tool that supports inking, importing/exporting pdfs seamlessly, and allows the handwriting to be searchable. Onenote is the closest app right now, but printing a large number of pages into Onenote is not ideal, it becomes convoluted and its slow to flip through notes. Exporting from onenote back to pdf does not work very well also.

**Comment by Haze-sh** [comment]
I would like to have the option of using Tesseract, maybe with other plugin?

**Comment by atticus-sullivan** [comment]
> I would like to have the option of using Tesseract, maybe with other plugin?

Unfortunately on the first glimpse there is no lua wrapper/bindings for tesseract OCR (looked at https://github.com/tesseract-ocr/tesseract#for-developers -> https://tesseract-ocr.github.io/tessdoc/AddOns.html#tesseract-wrappers).
Then, working with such an external (optional?) dependency might be harder than simply send some http requests.

**Comment by Haze-sh** [comment]
@atticus-sullivan If writing a Lua wrapper for tesseract would make a plugin easy I can work on this on a different project.

**Comment by atticus-sullivan** [comment]
Yes, I/we'd need to check how to pass the strokes to tesseract and how to 'find'/call tesseract. But `getStrokes()` is already in the making with that and the tesseract bindings is should somehow be feasable to do this. (Or at least implement it and check if the outcome is good enough)
But would you write an issue for this so that we have a place for discussion but don't interfer with rolandlos project?

**Comment by Haze-sh** [comment]
@atticus-sullivan I would love us to discuss this [here](https://github.com/Haze-sh/lua-tesseract/issues/1) so the two projects don't interfere with each other.

**Comment by atticus-sullivan** [comment]
@Haze-sh did you delete that project or made it private? (I'm getting a 404 error and can't see the project on your GH profile page)

**Comment by Haze-sh** [comment]
@atticus-sullivan Somehow it went private, it is public now!

**Comment by GabrielOsvair** [comment]
Sorry for my ignorance, but how can I use this handwriting recognition plugin? I don't find it available to enable.

**Comment by rolandlo** [comment]
@GabrielOsvair It's not in the official version (yet). You can find the files in https://github.com/xournalpp/xournalpp/pull/2176/files. If you are on Linux and run Xournal++ version 1.1.3+dev you could just copy those 5 files into `.config/xournalpp/plugins/HandwritingRecognition/` and then activate the plugin from the plugin manager.
Note that the plugin is quite limited though. At most 150 strokes can be recognized at once, which is not much.

**Comment by GabrielOsvair** [comment]
@rolandlo I understood! No problems. I use windows, so I'll be waiting for the development, because I don't understand about it to be able to help. Good luck with the project!

**Comment by hmpthcs** [comment]
This is a really good idea. My use case would be a proof-of-concept **stylus based on screen keyboard**. It seems such a tool does not exist in the linux world. Perplexingly, not at all. Ultimately, a very sleeved-down xournalpp running with a plugin like this could fill that gap.

I am getting an error stating something about line 33 in recognition.lua being a null value. Using xournalpp 1.1.3. Perhaps my version is not correct? Or perhaps I did not install the luarocks stuff to the right directory?

Also, is there any possibility to modify this plugin to use a native handwriting recognition? There are likely some existing out there, but don't know about quality or compatibility.

**Comment by rolandlo** [comment]
> This is a really good idea. My use case would be a proof-of-concept **stylus based on screen keyboard**. It seems such a tool does not exist in the linux world. Perplexingly, not at all. Ultimately, a very sleeved-down xournalpp running with a plugin like this could fill that gap.

Is there such an application on Windows or MacOS? If so, can you provide a link?

> 
> I am getting an error stating something about line 33 in recognition.lua being a null value. Using xournalpp 1.1.3. Perhaps my version is not correct? Or perhaps I did not install the luarocks stuff to the right directory?

You should install the [nightly build](https://github.com/xournalpp/xournalpp/releases/tag/nightly) in order to use the plugin.

**Comment by avidseeker** [comment]
https://github.com/xournalpp/xournalpp/discussions/5306

Not OCR, but simple search functionality for text.

**Comment by LHSequalsRHS** [comment]
Making the handwritten notes has another benefit of using "Semantic Ink". This feature is in Wacom Notes, like the one described in this tutorial - [Wacom Notes How to use Semantic Ink and Ink to Text](https://www.youtube.com/watch?v=Kn1Pucj7Ur0)

**Comment by Emily511511** [comment]
The most important feature without hesitation that would improve this app tenfolds.
With reliable, smooth, offline OCR as you write, this would easily be the best note taking app

**Comment by PellelNitram** [comment]
We are making progress with a plugin (called [xournalpp_htr](https://github.com/PellelNitram/xournalpp_htr)) that will bring recognition of handwritten text to Xournal++, see [this most recent devlog](https://youtu.be/yRWZeH2lsng) on how we are progressing :-) I'll keep you posted about future progress!
